shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureWrapper.java

```diff
@@ -0,0 +1,67 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.renderer;
+
+import android.graphics.SurfaceTexture;
+import androidx.annotation.NonNull;
+
+/**
+ * A wrapper for a SurfaceTexture that tracks whether the texture has been released.
+ *
+ * <p>The engine calls {@code SurfaceTexture.release} on the platform thread, but {@code
+ * updateTexImage} is called on the raster thread. This wrapper will prevent {@code updateTexImage}
+ * calls on an abandoned texture.
+ */
+public class SurfaceTextureWrapper {
+  private SurfaceTexture surfaceTexture;
+  private boolean released;
+
+  public SurfaceTextureWrapper(@NonNull SurfaceTexture surfaceTexture) {
+    this.surfaceTexture = surfaceTexture;
+    this.released = false;
+  }
+
+  @NonNull
+  public SurfaceTexture surfaceTexture() {
+    return surfaceTexture;
+  }
+
+  // Called by native.
+  @SuppressWarnings("unused")
+  public void updateTexImage() {
+    synchronized (this) {
+      if (!released) {
+        surfaceTexture.updateTexImage();
+      }
+    }
+  }
+
+  public void release() {
+    synchronized (this) {
+      if (!released) {
+        surfaceTexture.release();
+        released = true;
+      }
+    }
+  }
+
+  // Called by native.
+  @SuppressWarnings("unused")
+  public void attachToGLContext(int texName) {
+    surfaceTexture.attachToGLContext(texName);
+  }
+
+  // Called by native.
+  @SuppressWarnings("unused")
+  public void detachFromGLContext() {
+    surfaceTexture.detachFromGLContext();
+  }
+
+  // Called by native.
+  @SuppressWarnings("unused")
+  public void getTransformMatrix(float[] mtx) {
+    surfaceTexture.getTransformMatrix(mtx);
+  }
+}
```

Comment:
what about some unit tests for this class?

Reply:
The class doesn't contain any logic - it just forwards calls to `SurfaceTexture` with synchronization on the method called on the raster thread.  The overall Java/JNI interaction will be verified by end-to-end tests.

Reply:
+1